src/qpx_scan.cpp

```diff
@@ -0,0 +1,470 @@
+    if ((dev->media.type & DISC_CD) &&
+        (dev->capabilities & CAP_TEST_WRITE_CD)) {
+      printf("Turning TestWrite (CD) %s\n", simul ? "ON" : "OFF");
+      WT_simul = simul;
+      return 0;
+    } else if ((dev->media.type & DISC_DVDminus) &&
+               (dev->capabilities & CAP_TEST_WRITE_DVD)) {
+      printf("Turning TestWrite (DVD) %s\n", simul ? "ON" : "OFF");
+      WT_simul = simul;
+      return 0;
+    } else {
+      WT_simul = 0;
+      if (simul) {
+        if ((dev->media.type & DISC_DVDRAM)) {
+          printf("DVD-RAM media doesn't support TestWrite!\n");
+        } else {
+          printf("Drive doesn't support TestWrite on this media!\n");
+        }
+        return 1;
+      } else {
+        return 0;
+      }
+    }
+  }
+}
+
+void qscanner::stop() {
+  stop_req = 1;
+  if (writer)
+    writer->stop();
+};
+
+void qscanner::stat() { stat_req = 1; };
+
+int qscanner::run(char *test) {
+  int r = -1;
+  stop_req = 0;
+  stat_req = 0;
+  if (!dev->media.type) {
+    printf("No media detected!\n");
+    return 1;
+  }
+  if (!(dev->media.type & (DISC_CD | DISC_DVD | DISC_BD))) {
+    printf("Unsupported media!\n");
+    return 2;
+  }
+  //	set_speed(dev,speed);
+  if (!strcmp(test, "rt")) {
+    if (lba_end < 0 || lba_end > dev->media.capacity)
+      lba_end = dev->media.capacity - 1;
+    r = run_rd_transfer();
+  } else if (!strcmp(test, "wt")) {
+    if (lba_end < 0 || lba_end > dev->media.capacity_total)
+      lba_end = dev->media.capacity_total - 1;
+    r = run_wr_transfer();
+  } else if (!strcmp(test, "errc")) {
+    if (lba_end < 0 || lba_end > dev->media.capacity)
+      lba_end = dev->media.capacity - 1;
+    if (dev->media.type & DISC_CD) {
+      r = run_cd_errc();
+    } else if (dev->media.type & DISC_DVD) {
+      r = run_dvd_errc();
+    } else if (dev->media.type & DISC_BD) {
+      r = run_bd_errc();
+    }
+  } else if (!strcmp(test, "jb")) {
+    if (lba_end < 0 || lba_end > dev->media.capacity)
+      lba_end = dev->media.capacity - 1;
+    if (dev->media.type & DISC_CD) {
+      r = run_cd_jb();
+    } else if (dev->media.type & DISC_DVD) {
+      r = run_dvd_jb();
+    }
+  } else if (!strcmp(test, "ft")) {
+    if (lba_end < 0 || lba_end > dev->media.capacity_total)
+      lba_end = dev->media.capacity_total - 1;
+    r = run_fete();
+  } else if (!strcmp(test, "ta")) {
+    if (dev->media.type & DISC_CD) {
+      r = run_cd_ta();
+    } else if (dev->media.type & DISC_DVD) {
+      r = run_dvd_ta();
+    }
+  }
+  return r;
+}
+
+int qscanner::check_test(unsigned int test) {
+  if (!attached)
+    return -1;
+  return plugin->check_test(test);
+}
+
+int qscanner::errc_data() {
+  if (!attached)
+    return -1;
+  return plugin->errc_data();
+}
+
+int *qscanner::get_test_speeds(unsigned int test) {
+  if (!attached)
+    return NULL;
+  return plugin->get_test_speeds(test);
+}
+
+int qscanner::plugins_probe(bool test, bool probe_enable) {
+  char *pname;
+  char *ppath;
+  DIR *dir;
```

Comment:
Codacy found an issue: [The scope of the variable 'dir' can be reduced.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)